NBitcoin/BIP158/BitStream.cs

```diff
@@ -4,62 +4,165 @@
 namespace NBitcoin
 {
 	/// <summary> Provides a view of an array of bits as a stream of bits. </summary>
-	internal class BitStream
+	class BitStream
 	{
-		private readonly FastBitArray _buffer;
-		private int _position;
+		private byte[] _buffer;
+		private int _remainCount;
 
-		internal int Position
+		public BitStream()
+			: this(new byte[0])
 		{
-			get => _position;
-			set => _position = value;
 		}
 
-		public BitStream(FastBitArray bitArray)
+		public BitStream(byte[] buffer)
 		{
-			_buffer = bitArray;
-			_position = 0;
+			var newBuffer = new byte[buffer.Length];
+			Buffer.BlockCopy(buffer, 0, newBuffer, 0, buffer.Length);
+			_buffer = newBuffer;
+			_remainCount = buffer.Length == 0 ? 0 : 8;
 		}
 
-		public bool ReadBit()
+		private void AddZeroByte()
 		{
-			return _buffer[_position++];
+			var newBuffer = new byte[_buffer.Length + 1];
+			Buffer.BlockCopy(_buffer, 0, newBuffer, 0, _buffer.Length);
+			_buffer = newBuffer;
 		}
 
-		public ulong ReadBits(int count)
+		private void EnsureCapacity()
 		{
-			var ret = _buffer.GetBits(_position, count);
-			_position += count;
-			return ret;
+			if (_remainCount == 0)
+			{
+				AddZeroByte();
+				_remainCount = 8;
+			}
 		}
 
 		public void WriteBit(bool bit)
 		{
-			if (_buffer.Length == _position)
+			EnsureCapacity();
+			if (bit)
+			{
+				var lastIndex = _buffer.Length - 1;
+				_buffer[lastIndex] |= (byte)(1 << (_remainCount - 1));
+			}
+			_remainCount--;
+		}
+
+        public void WriteBits(ulong data, byte count)
+        {
+			data <<= (64 - count);
+			while(count >= 8)
+			{
+				var b = (byte)(data >> (64 - 8));
+				WriteByte(b);
+				data <<= 8;
+				count -= 8;
+			}
+
+			while(count > 0)
+			{
+				var bit = data >> (64 - 1);
+				WriteBit(bit == 1);
+				data <<= 1;
+				count--;
+			}
+		}
+
+		public void WriteByte(byte b)
+		{
+			EnsureCapacity();
+
+			var lastIndex = _buffer.Length - 1;
+			_buffer[lastIndex] |= (byte)(b >> (8 - _remainCount));
+
+			AddZeroByte();
+			_buffer[lastIndex + 1] = (byte)(b << _remainCount);
+		}
+
+		public bool ReadBit()
+		{
+			if (_buffer.Length == 0)
+				throw new InvalidOperationException("The stream is empty");
```

Comment:
Can you add punctuation, like this: (`"The stream is empty."`) so to comply with Microsoft's best practices? https://docs.microsoft.com/en-us/dotnet/standard/exceptions/best-practices-for-exceptions#use-grammatically-correct-error-messages